### PR TITLE
Add mlir-trunk

### PR DESCRIPTION
**Pull request by Sameeranjoshi**

### CODE REVIEW

**Comment by Sameeranjoshi** [comment]
@partouf tried this random changes, probably I did them right. 
Let me know any pointers so I can speed up this process of adding support for mlir-trunk.
Thanks

**Review by jeremy-rifkin — COMMENTED** [review]
These changes look correct in principal. I'm not familiar with how everything works here but as long as an s3 tarball named `mlir-trunk` exists it's my understanding this should work.

File: bin/yaml/cpp.yaml [hunk 1]
```diff
@@ -758,6 +758,7 @@ compilers:
         - 14.0.0
         - 14.0.5
         - 16.0.0
+        - trunk
```
Comment (mattgodbolt) [inline bin/yaml/cpp.yaml:761]: in this way it'll only get installed once: the `trunk` stuff needs to be done slightly differently in order to not fall afoul of our "don't reinstall things". Plus trunk builds are actually dated... I'll fix up in a separate change; but thanks!

**Comment by mattgodbolt** [comment]
Getting there - sorry for the delay. In investigating this, as far as I can tell the "mlir" library is installing the exact same thing as the compiler...so 

```
Jun/02 11:15 admin-node~/infra (main|✔) 1.4s $ ls /opt/compiler-explorer/mlir-14.0.0/
bin  include  lib  share
```

is the same as
```
Jun/02 12:49 admin-node~/infra (main|✔) $ ls /opt/compiler-explorer/libs/mlir/14.0.0/
bin  include  lib  share
```

I don't really understand why - what is the "mlir" library? On LLVM we have a separate build for LLVM the library vs clang the compiler. Do you know what this might be for MLIR? Right now it seems just a waste of disk space to install the same thing in two places

**Comment by joker-eph** [comment]
MLIR is similar to LLVM: it is a library but also comes with tools (LLVM has `opt` and MLIR has `mlir-opt`). There should be no reason to duplicate the tree, it should be aligned with whatever is done for LLVM I think?

**Pull request merged**